types/wlr_virtual_pointer_v1.c

```diff
@@ -0,0 +1,291 @@
+		struct wl_resource *resource, uint32_t time, uint32_t button,
+		uint32_t state) {
+	if (button == 0) {
+		return;
+	}
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_button event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.button = button,
+		.state = state ? WLR_BUTTON_PRESSED : WLR_BUTTON_RELEASED
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.button, &event);
+}
+
+static void virtual_pointer_axis(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t source,
+		wl_fixed_t value) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_axis event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.source = source,
+		.delta = wl_fixed_to_double(value),
+		.delta_discrete = wl_fixed_to_int(value),
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.axis, &event);
+}
+
+static void virtual_pointer_frame(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.frame, wlr_dev->pointer);
+}
+
+static void virtual_pointer_axis_source(struct wl_client *client,
+		struct wl_resource *resource, uint32_t source) {
+	// Not supported at the moment
+}
+
+static void virtual_pointer_axis_stop(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t source) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_axis event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.source = source,
+		.delta = 0,
+		.delta_discrete = 0
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.axis, &event);
+}
+
+static void virtual_pointer_axis_discrete(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t source,
+		wl_fixed_t value, int32_t discrete) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_axis event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.source = source,
+		.delta = wl_fixed_to_double(value),
+		.delta_discrete = discrete
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.axis, &event);
+}
+
+static void virtual_pointer_destroy_resource(struct wl_resource *resource) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	wlr_signal_emit_safe(&pointer->events.destroy, pointer);
+	wl_list_remove(&pointer->link);
+	wlr_input_device_destroy(&pointer->input_device);
+	free(pointer);
```

Comment:
The compositor can destroy the device too. We should move this logic into `input_device_destroy` and make the resource inert from there.

More info about inert resources in `CONTRIBUTING`. Feel free to ask questions about all of this on IRC, too.